File: day6/day6.cpp
#include <bits/stdc++.h>
#define ll long long int
using namespace std;

vector<ll> parseline(string s) {
    vector<ll> l = {};
    size_t pos = s.find(' ');
    while (pos < s.length()) {
        size_t npos = s.find(' ', pos+1);
        if (s[pos+1] != ' ')
            l.push_back(stoll(s.substr(pos+1,npos)));
        pos = npos;
    }
    return l;
}

ll parseline2(string s) {
  vector<char> digits; 
  size_t pos = s.find(' ');
  while (pos < s.length()) {
    if (s[pos] != ' ') digits.push_back(s[pos]);
    pos++;
  }
  string r(digits.begin(), digits.end());
  return stoll(r);
}

ll nwins(ll t, ll d) {
  double sD = sqrt(t * t - 4 * d);
  double x1r = (t - sD) / 2 + .001;
  double x2r = (t + sD) / 2 - .001;
  ll x1 = ceil(x1r);
  ll x2 = floor(x2r);
  return x2 - x1 + 1;
}

int main() {
//    freopen("test.txt", "r", stdin);
    string times, dists;
    getline(cin, times);
    getline(cin, dists);

    vector<ll> ts = parseline(times);
    vector<ll> ds = parseline(dists);
   
    ll total = 1; 
    for (size_t i = 0; i < ts.size(); i++) {
        total *= nwins(ts[i],ds[i]); 
    }

    cout << total << endl;

    ll t = parseline2(times);
    ll d = parseline2(dists);

    cout << nwins(t, d) << endl; 
}

File: day6/day6.py
from utils import *
from math import ceil, floor, sqrt
from functools import reduce
import operator

def parse1(ls):
    times, dists = [[int(n) for n in l.split(" ")[1:] if n != ""] for l in ls]
    return (times,dists)

def parse2(ls):
    t, d = [int(l.replace(" ", "").split(":")[1]) for l in ls]
    return (t,d)
    
def quadratic(a,b,c):
    D = b * b - 4 * a * c
    if D < 0: return None, None
    sD = sqrt(D)
    return (-b - sD) / (2 * a), (-b + sD) / (2 * a)

def nwins(t,d):
    # distance traveled when button held for x seconds = x * (t - x)
    # to solve x * (t - x) > d, we want x in [0,t] so that x^2 - tx + d < 0
    x1, x2 = quadratic(1, -t, d)
    # assert x1 is not None       # there's always some way to be better
    eps = .000001               # need a little offset because we want strictly better 
    x1r, x2r = ceil(x1+eps),floor(x2-eps)
    return x2r - x1r + 1

def solve1(ls):
    ts, ds = parse1(ls)
    return reduce(operator.mul, [nwins(ts[i],ds[i]) for i in range(len(ts))])

def solve2(ls):
    t, d = parse2(ls)
    return nwins(t,d)

def solve(filename, expected=None):
    with open(filename) as f:
        ls = f.readlines()
    s = [solve1(ls), solve2(ls)]
    
    report(filename, s, expected)
    
if __name__ == "__main__":
    solve("test.txt", [288,71503])
    solve("input.txt")
